lib/cli/src/Cardano/CLI.hs

```diff
@@ -431,6 +447,64 @@ instance FromText CliWalletStyle where
 instance ToText CliWalletStyle where
     toText = toTextFromBoundedEnum SnakeLowerCase
 
+newtype DerivationPath = DerivationPath [DerivationIndex]
+    deriving (Show, Eq)
+
+instance FromText DerivationPath where
+    fromText x = DerivationPath <$> mapM fromText (T.splitOn "/" x)
+
+instance ToText DerivationPath where
+    toText (DerivationPath xs) = T.intercalate "/" $ map toText xs
+
+newtype DerivationIndex = DerivationIndex { unDerivationIndex :: Word32 }
+    deriving (Show, Eq)
+    deriving newtype (Bounded, Enum)
+
+instance FromText DerivationIndex where
+    fromText "" = Left $ TextDecodingError
+        "An empty string is not a derivation index!"
+    fromText x = do
+       -- NOTE: T.takeEnd will not throw, but may return "".
+       if T.takeEnd 1 x == "\'"
+       then do
+           let x' = T.dropEnd 1 x
+           parseHardenedIndex x'
+       else parseSoftIndex x
+      where
+        parseWord = left (const err) .
+            fmap fromIntegral . fromText @Int
+          where
+            err = TextDecodingError $
+                "\"" ++ T.unpack x ++ "\" is not a number."
+
+        mkDerivationIndex :: Int -> Either TextDecodingError DerivationIndex
+        mkDerivationIndex = inv >=> (return . DerivationIndex . toEnum . fromEnum)
+
+        parseSoftIndex =
+            parseWord >=> mkDerivationIndex
+
+        parseHardenedIndex =
+            parseWord
+            >=> (return . (+ 0x80000000))
```

Comment:
Should probably leverage the existing type
```Haskell
newtype Index (derivationType :: DerivationType) (level :: Depth) = Index
```
to avoid hard-coding `0x80000000` again.

Reply:
However, I'm not convinced that we need as much flexibility as the `Index` type provides.

For now I resorted to defining
```Haskell
firstHardenedIndex :: Word32
firstHardenedIndex = getIndex $ minBound @(Index 'Hardened 'AddressK)
```